### ISSUE TITLE
no-undef collides with 'with' keyword because it ignores scope

### ISSUE THREAD

**Issue by StephanBijzitter** [eslint/eslint#1354]
``` javascript
with (Cookies.backing) {
    var result = typeof filter === 'string'
        ? get(filter) : matches(filter);
}
```

`warning  'matches' is not defined  no-undef`

This warning is thrown because ESLint seems to think the current scope is still `Cookie` (which has `get` defined but not `matches`). However, `Cookies.backing` is the scope and that object does have both the `get` and `matches` functions defined, which means the warning is incorrect.
